# Homicide Rates per County

In [2]:
import pandas as pd 

In [3]:
hom_by_cty = pd.read_csv('cty_homicides_2017.txt', delimiter='\t')
hom_by_cty.head(1).T

0
Notes                       NaN
County       Autauga County, AL
County Code                1001
Deaths               Suppressed
Population                55504
Crude Rate           Suppressed

In [4]:
hom_by_cty.drop(['Notes'], inplace=True, axis=1)

In [5]:
#this is homicide rate per county where there is enough data 
hom_by_cty.head()

County  County Code      Deaths Population  Crude Rate
0  Autauga County, AL       1001.0  Suppressed      55504  Suppressed
1  Baldwin County, AL       1003.0  Suppressed     212628  Suppressed
2  Barbour County, AL       1005.0  Suppressed      25270  Suppressed
3     Bibb County, AL       1007.0  Suppressed      22668  Suppressed
4   Blount County, AL       1009.0  Suppressed      58013  Suppressed

# County Features 

We tried using the ACS 1 year data through Cenpy and the API and consistently got ~820 to 840 counties. 

I tried it from the API directly and from cenpy, with different years for each, and it always returned counties within that range. And, not all states had counties in there, so it wasn't a random sample (remember, Arizona had zero...) 

#### However, using the 5 year data does get us more county information! 

In [6]:
import cenpy as c
#find table that we want to query 
available = c.explorer.available()
acs_df = available[available['title'].str.contains('ACS') == True]
acs_df = acs_df[acs_df['vintage'] == 2017]
acs_df

/Users/tjjj/anaconda3/lib/python3.7/site-packages/libpysal/weights/util.py:19: UserWarning: geopandas not available. Some functionality will be disabled.
  warn('geopandas not available. Some functionality will be disabled.')


variable_cell_limit                                    title  \
ACSDT1Y2017                  NaN               ACS 1-Year Detailed Tables   
ACSCP1Y2017                  NaN           ACS 1-Year Comparison Profiles   
ACSDP1Y2017                  NaN                 ACS 1-Year Data Profiles   
ACSSPP1Y2017                 NaN  ACS 1-Year Selected Population Profiles   
ACSST1Y2017                  NaN                ACS 1-Year Subject Tables   
ACSDT5Y2017                  NaN               ACS 5-Year Detailed Tables   
ACSCP5Y2017                  NaN           ACS 5-Year Comparison Profiles   
ACSDP5Y2017                  NaN                 ACS 5-Year Data Profiles   
ACSST5Y2017                  NaN                ACS 5-Year Subject Tables   
ACSSE2017                    NaN        ACS 1-Year Supplemental Estimates   

                  temporal spatial           publisher programCode  \
ACSDT1Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSCP1Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSDP1Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSSPP1Y2017  unidentified     NaN  U.S. Census Bureau     006:004   
ACSST1Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSDT5Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSCP5Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSDP5Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSST5Y2017   unidentified     NaN  U.S. Census Bureau     006:004   
ACSSE2017     unidentified     NaN  U.S. Census Bureau     006:004   

                           modified keyword  \
ACSDT1Y2017   2018-09-13 00:00:00.0      ()   
ACSCP1Y2017   2018-09-13 00:00:00.0      ()   
ACSDP1Y2017   2018-09-13 00:00:00.0      ()   
ACSSPP1Y2017  2018-09-17 00:00:00.0      ()   
ACSST1Y2017   2018-09-13 00:00:00.0      ()   
ACSDT5Y2017   2018-08-21 07:11:43.0      ()   
ACSCP5Y2017   2018-10-19 00:00:00.0      ()   
ACSDP5Y2017   2018-10-19 00:00:00.0      ()   
ACSST5Y2017   2018-10-19 00:00:00.0      ()   
ACSSE2017     2018-10-18 00:00:00.0      ()   

                                                   distribution  \
ACSDT1Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSCP1Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSDP1Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSSPP1Y2017  {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSST1Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSDT5Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSCP5Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSDP5Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSST5Y2017   {'@type': 'dcat:Distribution', 'accessURL': 'h...   
ACSSE2017     {'@type': 'dcat:Distribution', 'accessURL': 'h...   

                                                    description  ... c_isCube  \
ACSDT1Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSCP1Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSDP1Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSSPP1Y2017  Selected Population Profiles provide broad soc...  ...     True   
ACSST1Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSDT5Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSCP5Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSDP5Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSST5Y2017   The American Community Survey (ACS) is an ongo...  ...     True   
ACSSE2017     The American Community Survey (ACS) is a natio...  ...     True   

              c_isAvailable c_isAggregate  \
ACSDT1Y2017            True          True   
ACSCP1Y2017            True          True   
ACSDP1Y2017            True          True   
ACSSPP1Y2017           True          True   
ACSST1Y2017          

Based on this website and other research, we want to use ACSDP5Y2017.

In [7]:
c.explorer.explain('ACSDP5Y2017')

{'ACS 5-Year Data Profiles': 'The American Community Survey (ACS) is an ongoing survey that provides data every year -- giving communities the current information they need to plan investments and services. The ACS covers a broad range of topics about social, economic, demographic, and housing characteristics of the U.S. population. The data profiles include the following geographies: nation, all states (including DC and Puerto Rico), all metropolitan areas, all congressional districts, all counties, all places and all tracts. Data profiles contain broad social, economic, housing, and demographic information. The data are presented as both counts and percentages. There are over 2,400 variables in this dataset.'}

In [8]:
con = c.base.Connection('ACSDP5Y2017')
g_unit = 'county:*'

In [9]:
#create education features list from https://api.census.gov/data/2017/acs/acs5/profile/variables.html
cols_edu = [] 
for n in range(58,67): 
    var_name = 'DP02_00'+str(n)+'PE'
    cols_edu.append(var_name)

#create internet features list
cols_internet = []
for n in range(150,152):
    var_name = 'DP02_0'+str(n)+'PE'
    cols_internet.append(var_name)

#create row names for joining and EDA
cols_req = ['NAME']


In [10]:
cols = cols_req + cols_edu + cols_internet 
cols

['NAME',
 'DP02_0058PE',
 'DP02_0059PE',
 'DP02_0060PE',
 'DP02_0061PE',
 'DP02_0062PE',
 'DP02_0063PE',
 'DP02_0064PE',
 'DP02_0065PE',
 'DP02_0066PE',
 'DP02_0150PE',
 'DP02_0151PE']

In [11]:
counties_df = con.query(cols=cols, geo_unit=g_unit)
counties_df.head()

NAME DP02_0058PE DP02_0059PE DP02_0060PE DP02_0061PE  \
0    Pickens County, Alabama       14241         6.2        13.9        37.8   
1     Sumter County, Alabama        8244         4.7        12.6        39.7   
2  Jefferson County, Alabama      447048         3.0         7.6        26.8   
3    Choctaw County, Alabama        9449         6.6        13.3        39.9   
4   Franklin County, Alabama       20734        11.8        11.9        37.5   

  DP02_0062PE DP02_0063PE DP02_0064PE DP02_0065PE DP02_0066PE DP02_0150PE  \
0        22.8         7.5         8.8         3.0        79.8        7620   
1        17.7         7.0        10.6         7.6        82.7        5073   
2        22.6         8.1        19.4        12.5        89.4      261390   
3        19.5         9.1         7.9         3.8        80.1        5463   
4        18.0         7.5         8.3         5.1        76.4       11533   

  DP02_0151PE state county  
0        71.0    01    107  
1        64.8    01    119  
2        84.4    01    073  
3        70.4    01    023  
4        74.2    01    059

P.S. I got the API to return the same information except it wasn't as easy to work with especially because the strings are long and the column names don't autopopulate with the state and county names. Also of note: When I was getting errors in the wrapper they didn't have details, but the errors through the browser version of the API had details.

# Compare the target to the features


In [12]:
#how many counties are in the target set? 
hom_by_cty.County.nunique()
#note that some are suppressed 

3147

In [13]:
#how many counties are in the target set? 
print(len(counties_df))
#note that these include Puerto Rico

3220


Pretty close! Maybe we can limit by population first (on homicides df) then join the data on the county name. 

In [14]:
import numpy as np
hom_by_cty.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3213 entries, 0 to 3212
Data columns (total 5 columns):
County         3147 non-null object
County Code    3147 non-null float64
Deaths         3148 non-null object
Population     3148 non-null object
Crude Rate     3148 non-null object
dtypes: float64(1), object(4)
memory usage: 125.6+ KB


In [15]:
#convert columns to usable datatypes
hom = hom_by_cty
hom.head()

County  County Code      Deaths Population  Crude Rate
0  Autauga County, AL       1001.0  Suppressed      55504  Suppressed
1  Baldwin County, AL       1003.0  Suppressed     212628  Suppressed
2  Barbour County, AL       1005.0  Suppressed      25270  Suppressed
3     Bibb County, AL       1007.0  Suppressed      22668  Suppressed
4   Blount County, AL       1009.0  Suppressed      58013  Suppressed

In [16]:
len(hom[hom['Crude Rate']!='Suppressed'])

414

Only 414 rows have usable data. We're technically supposed to have 1000 rows, which we start out with, before we remove the ones that are suppressed. What if we looked at suicides instead? 

# Suicide Rates by County
Just out of curiosity, is there more data to work with around suicides?  

In [17]:
suicides = pd.read_csv('suicides.txt', delimiter='\t').drop(['Notes'], axis=1)

print(len(suicides))
suicides.head()
#we don't have all counties here

1084


County  County Code  Deaths  Population  Crude Rate
0    Autauga County, AL       1001.0    12.0     55504.0  Unreliable
1    Baldwin County, AL       1003.0    39.0    212628.0        1.83
2     Blount County, AL       1009.0    11.0     58013.0  Unreliable
3    Calhoun County, AL       1015.0    29.0    114728.0        2.53
4  Covington County, AL       1039.0    12.0     37092.0  Unreliable

In [18]:
suicides.drop(suicides[suicides['Crude Rate'] == 'Unreliable'].index, inplace=True)

In [19]:
print(len(suicides))
suicides.head()

607


County  County Code  Deaths  Population Crude Rate
1     Baldwin County, AL       1003.0    39.0    212628.0       1.83
3     Calhoun County, AL       1015.0    29.0    114728.0       2.53
5     Cullman County, AL       1043.0    21.0     82755.0       2.54
6      DeKalb County, AL       1049.0    20.0     71617.0       2.79
12  Jefferson County, AL       1073.0   109.0    659197.0       1.65

In [20]:
suicides.drop(suicides[suicides['Crude Rate'] == 'Suppressed'].index, inplace=True)
suicides.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 607 entries, 1 to 1083
Data columns (total 5 columns):
County         536 non-null object
County Code    536 non-null float64
Deaths         537 non-null float64
Population     537 non-null float64
Crude Rate     537 non-null object
dtypes: float64(3), object(2)
memory usage: 28.5+ KB


# To Alex from Mia - where do we go from here? 
So we know how to get the data from the CDC on deaths/other health things, and we know how to get data from the Census. That's pretty powerful, even if we aren't sure if we have "enough" data for the mod 3 project. I don't want to scrap our idea entirely and use a different data set, because I think this is really cool and just worked on it a bunch, lol. What I would consider doing from here is: 
1. Get permission from an instructor to target homicide rates OR suicide rates even though there aren't technically over 1000 of them. 
2. Target both of those and compare if we have time. 
3. Choose a new death or injury variable that is more "common" than suicide and homicide like cardiac disease and use that if it returns more than 1000 rows. 
3. Choose a new target variable from the actual Census data itself, now that we know how to query it.

Let me know your thoughts! Text me when/if you work on this at all. :) 

# Homicide Rates 2013-2017 

In [36]:
homicides = pd.read_csv('2013-2017 Homicide Information.txt', delimiter='\t')
homicides = homicides.drop(['Notes'], axis=1)

In [65]:
homicides = homicides.drop(homicides[homicides['Crude Rate']=='Suppressed'].index)

In [66]:
homicides_reliable = homicides.drop(homicides[homicides['Crude Rate']=='Unreliable'].index)
homicides_reliable = homicides_reliable.drop(homicides_reliable[homicides_reliable['Crude Rate'].isna()].index)

In [67]:
homicides_reliable = homicides_reliable.drop(homicides_reliable[homicides_reliable['Deaths']=='Missing'].index)
homicides_reliable = homicides_reliable.drop(homicides_reliable[homicides_reliable['Deaths'].isna()].index)

In [69]:
homicides_reliable['Deaths'].nunique()

207

In [70]:
homicides_reliable.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 633 entries, 1 to 3147
Data columns (total 6 columns):
County               632 non-null object
County Code          632 non-null float64
Deaths               633 non-null object
Population           633 non-null object
Crude Rate           633 non-null object
% of Total Deaths    633 non-null object
dtypes: float64(1), object(5)
memory usage: 34.6+ KB


In [103]:
od = pd.read_csv('2013-2017 Overdose Information.txt', delimiter='\t')
od = od.drop(['Notes'], axis=1)

od = od.drop(od[od['Crude Rate']=='Suppressed'].index)

od_reliable = od_reliable.drop(od_reliable[od_reliable['Crude Rate']=='Unreliable'].index)
od_reliable = od_reliable.drop(od_reliable[od_reliable['Deaths']=='Missing'].index)
od_reliable = od_reliable.drop(od_reliable[od_reliable['Deaths'].isna()].index)
od_reliable = od_reliable.drop(od_reliable[od_reliable['County Code'].isna()].index)
od_reliable.head()

County  County Code Deaths Population Crude Rate  \
0  Autauga County, AL       1001.0     47     276908    16.9731   
1  Baldwin County, AL       1003.0    253    1020551    24.7905   
3     Bibb County, AL       1007.0     27     112912    23.9124   
4   Blount County, AL       1009.0     88     288981    30.4518   
7  Calhoun County, AL       1015.0    132     577611    22.8528   

  % of Total Deaths  
0           0.0103%  
1           0.0554%  
3           0.0059%  
4           0.0193%  
7           0.0289%

In [104]:
od_reliable.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2016 entries, 0 to 3144
Data columns (total 6 columns):
County               2016 non-null object
County Code          2016 non-null float64
Deaths               2016 non-null object
Population           2016 non-null object
Crude Rate           2016 non-null object
% of Total Deaths    2016 non-null object
dtypes: float64(1), object(5)
memory usage: 110.2+ KB


In [113]:
od_reliable['County Code'] = od_reliable['County Code'].astype(int)
od_reliable['Deaths'] = od_reliable['Deaths'].astype(int)
od_reliable['Population'] = od_reliable['Population'].astype(int)
od_reliable['Crude Rate'] = od_reliable['Crude Rate'].astype(float)

In [114]:
od_reliable.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2016 entries, 0 to 3144
Data columns (total 6 columns):
County               2016 non-null object
County Code          2016 non-null int64
Deaths               2016 non-null int64
Population           2016 non-null int64
Crude Rate           2016 non-null float64
% of Total Deaths    2016 non-null object
dtypes: float64(1), int64(3), object(2)
memory usage: 110.2+ KB


In [135]:
od_reliable['% of Total Deaths'] = od_reliable['% of Total Deaths'].str.replace("%", "")
od_reliable['% of Total Deaths'] = od_reliable['% of Total Deaths'].astype(float)

In [136]:
od_df = od_reliable

In [137]:
od_df.head()

County  County Code  Deaths  Population  Crude Rate  \
0  Autauga County, AL         1001      47      276908     16.9731   
1  Baldwin County, AL         1003     253     1020551     24.7905   
3     Bibb County, AL         1007      27      112912     23.9124   
4   Blount County, AL         1009      88      288981     30.4518   
7  Calhoun County, AL         1015     132      577611     22.8528   

   % of Total Deaths  
0             0.0103  
1             0.0554  
3             0.0059  
4             0.0193  
7             0.0289

In [138]:
od_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2016 entries, 0 to 3144
Data columns (total 6 columns):
County               2016 non-null object
County Code          2016 non-null int64
Deaths               2016 non-null int64
Population           2016 non-null int64
Crude Rate           2016 non-null float64
% of Total Deaths    2016 non-null float64
dtypes: float64(2), int64(3), object(1)
memory usage: 110.2+ KB
